### ISSUE TITLE
Support LoRA for nn.Embeddings

### ISSUE THREAD

**Issue by Andrei-Aksionov** [Lightning-AI/lit-gpt#268]
Hello @carmocca 👋 

How do you think: should be `LoRA` for `nn.Embeddings` supported by this repo? That's basically the last weight matrix that is not supported.
And the [code](https://github.com/microsoft/LoRA/blob/main/loralib/layers.py#L32) is in the orig repo anyway.

Just want to clarify it before spending time on a PR.

**Comment by carmocca** [comment]
Do you know of any papers or projects indicating that this is important?
Without strong motivation, I wouldn't add this option. Of course, users are free to fork and implement this themselves if they need it. But we don't need to implement every feature if they aren't generally used.

**Comment by Andrei-Aksionov** [comment]
That is the problem: I couldn't find that anyone mention it. Hoped maybe you've heard something about it.
Anyway, glad that I've asked before start coding.

I also created an issue in the QLoRA repo with this question: https://github.com/artidoro/qlora/issues/210
If the answer is negative - I'll close this issue.

**Comment by apachemycat** [comment]
> Do you know of any papers or projects indicating that this is important? Without strong motivation, I wouldn't add this option. Of course, users are free to fork and implement this themselves if they need it. But we don't need to implement every feature if they aren't generally used.

 https://github.com/artidoro/qlora/issues/214 
**A more general fix would be adding LoRA layers to the embeddings or allowing only the new embeddings to be trained. The LoRA layer for embeddings might not work as well on the output projection layer (mapping back to the vocabulary before softmax).**

**Comment by Andrei-Aksionov** [comment]
@apachemycat Thanks for the message 👍

The issue as I see might happen only if new new tokens are added (no sure that this repo is doing it) or when we do [weight tying](https://paperswithcode.com/method/weight-tying) - when `embeddings` and `lm_head` have the same weight matrix (this repo definitely is not doing it).

So indeed, one of the reasons why LoRA Embeddings might be helpful - when embeddings resizing functionality is added to the repo.

**Comment by Andrei-Aksionov** [comment]
So the [response](https://github.com/artidoro/qlora/issues/210#issuecomment-1645277393) was:
> We have not tried this in our experiments but I think it would be quite interesting! LoRA layer for embedding would need to be supported by PEFT and we did not implement it. This might be useful when training on a new language or domain.

Which means that it's basically (for now) a nice to have functionality.
@carmocca  Postpone the implementation for now?

**Comment by carmocca** [comment]
I'd say yes until further evidence is presented. The issue can be left open
